### ISSUE TITLE
Is there any advice to optimize the point?

### ISSUE THREAD

**Issue by shakarelic** [moquette-io/moquette#204]
when max>5000 :error
Jmeter test 5min,200 Concurrent ->Subscribe and publish and Unsubscribe
moquette 
Label   #Samples Average Median 90%Line 95%Line 99%Line Min Max Error% Throughput KB/Sec
200         !43969    !1326 !278      !3517 !3790   !3996      !2     !5012  0.1%      145.6    18.34

mosquitto
Label #Samples Average Median 90%Line 95%Line 99%Line Min Max Error% Throughput KB/Sec
200   !1984108     !29    !1    !1  !1  !2  !0  !5475       !0.006% 6513   816.7

The gap is so big , 
Is there any advice to optimize the point?

**Comment by andsel** [comment]
The answer is yes. But I dont' know the test you ran, could you explain it?
 Andrea

**Comment by shakarelic** [comment]
Jmeter test : Run for (five minutes and 200 Concurrent thread )

one thread begin : Subscribe and publish and Unsubscribe : end
Samples mean : run count request
Average mean : Average response time(milliseconds)

**Comment by andsel** [comment]
I mean, was the test in local host or not. Was the first JVM run or these
are the results after warmed up the JVM. Give some context about the test.
The connections was done using a ramp or not? It's possible to share the
JMeter project file so that I musn't reinvent it, or better generat e a
Groovy/Java script that could do the same?

 Andrea

**Comment by shakarelic** [comment]
both test with local host and (or not)
no first jvm,and test many times ,you can know 
mosquitto Average response time(20  milliseconds)
moquette Average response time(1326  milliseconds)
sorry my english is poor. i think you can understand my means

**Comment by andsel** [comment]
Ok @shakarelic, there's no poblems with english, we need to understand.
Ususally in moquette, I write performance tests in Groovy using HdrHistogram, I need to recrete your test fixture, could you please answer some questions?
1) which version of moquette are you using?
2) OS?
3) you test was a ramp of thread or you bulk started 200 threads doing SUB (to which topic?) PUB UNSUB?
4) could you share the project definition file created by JMeter (so that I can import in my instance)?
 Thanks
   Andrea

**Comment by shakarelic** [comment]
you can test : 
one client-- and  more thread
Pseudo code --

```
for(long i=0L;i<2000000L;i++){
    closePool.execute({
        new Runnable(){
            public void run(){
                SettableFuture<String>set=SettableFuture.create();
                sub('test'+i,0);
                pub('test'+i,0);
                set.get(10000,TimeUnit.MILLISECONDS);
                unsub('test'+1);
            }
        }
    })
}
```

**Comment by andsel** [comment]
Sorry but this pseudo code is not much usefull.
- You submit 2M tasks to a pool and you don't say the dimension of Threads used in the pool.
- Use SettableFuture that  is not clear who sets? What's this SettableFuture, not standard JDK stuff

**Comment by andsel** [comment]
This is due to the fact that on publish was invoked (to do have better performance on heavy publish storm on some client session) just write and not writeAndFlush. Fixed with cf67d5ac524108173bc46014919d8fbd40b7c056
